secure-api-gateway-ig-extensions/src/main/java/com/forgerock/sapi/gateway/dcr/utils/DCRUtils.java

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright © 2020-2022 ForgeRock AS (obst@forgerock.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.forgerock.sapi.gateway.dcr.utils;
+
+import javax.validation.constraints.NotNull;
+
+import org.apache.ivy.util.StringUtils;
+import org.forgerock.json.jose.common.JwtReconstruction;
+import org.forgerock.json.jose.exceptions.JwtReconstructionException;
+import org.forgerock.json.jose.jws.SignedJwt;
+import org.forgerock.json.jose.jwt.JwtClaimsSet;
+
+import com.forgerock.sapi.gateway.dcr.request.DCRSignatureValidationException;
+import com.forgerock.sapi.gateway.dcr.request.DCRSignatureValidationException.ErrorCode;
+import com.forgerock.sapi.gateway.trusteddirectories.TrustedDirectory;
+import com.forgerock.sapi.gateway.trusteddirectories.TrustedDirectoryService;
+
+/**
+ * A utility class providing methods used by the dcr.request classes
+ */
+public class DCRUtils {
+
+    /**
+     * get the value 'iss' claim, which hold a value indicating the issuer of the JWT.
+     * @param jwtName the name of the jwt from which the issuer is being obtained. Used in the error_description
+     *                of the exception thrown if the issuer can't be obtained
+     * @param signedJwt the signed JWT from which to obtain the issuer
+     * @return a string containing the issuer
+     * @throws DCRSignatureValidationException if the JWT does not contain an issuer claim
+     */
+    @NotNull
+    public String getJwtIssuer(String jwtName, SignedJwt signedJwt) throws DCRSignatureValidationException {
```

Comment:
I prefer to pass a class into the constructor of dependent classes. I don't like statics as it make mocking/testing difficult.

Reply:
Passing dependencies via the constructor is definitely the way to go, however I don't see Utils classes as dependencies.

I'm not a big fan of these utils style classes but what you have here seems ok. It adheres to the convention of a utils class: stateless, immutable, side affect free, thread safe. Therefore there should be no need to ever mock this thing out.

For testing, you will need to provide valid test data / objects which these utils methods can operate on. I don't think this is a bad thing, I prefer to pass proper objects around in test rather than partially mocked stuff.

Otherwise, if you want to treat this as a true dependency, then the name needs some thought. Vaguely named classes tend to become dumping grounds for stuff, and as the DcrUtils object gets passed everywhere then it becomes a convenient way to fudge things.

Reply:
Some more thoughts about this issue, I like that you are trying to follow a DI approach and supply the dependency, the issue I have is that there is a Utils class idiom in java and this class is not following it.

On the other hand, I don't particularly like the static Utils class idiom so it's probably not worthwhile making DcrUtils follow this.

Either way, having such a class seems to me to violate the Single Responsibility principle so should really be avoided.

Reviewing the class again I propose the following:

1. com.forgerock.sapi.gateway.dcr.utils.DCRUtils#getSignedJwt , this is not DCR specific and would be useful whenever we need to decode a base64 string representation of a JWT, therefore it should be moved into a JwtDecoder class in package: com.forgerock.sapi.gateway.jws
2. The remaining methods appear to relate to extracting data from the Registration Request, therefore I think we should create a RegistrationRequest object and a RegistrationRequestDecoder (which builds the object from the HTTP Request). Then we can reject upfront badly formed request and provide instance methods which extract the data we need

I think this is a much better OOP approach.